### ISSUE TITLE
Support multiple IHostLifetime in Host class

### ISSUE THREAD

**Issue by chuanboz** [aspnet/Hosting#1401]
Currently in Host.cs the constructor injection only allow one single IHostLifetime to be used.

https://github.com/aspnet/Hosting/blob/edada4acc744c6257e48b4e474ee2371fa0260db/src/Microsoft.Extensions.Hosting/Internal/Host.cs#L24

The request is to change it to IEnumerable<IHostLifetime> so that multiple IHostLifetime could be registered each one dedicated to one function for host lifetime.

**Comment by Tratcher** [comment]
We considered a similar item (https://github.com/aspnet/Hosting/issues/1221) but decided against it. There weren't enough examples of different lifetimes that you would want to use together. 

Which types of lifetime would you want to use together?

**Comment by chuanboz** [comment]
Here is the scenario for Microsoft Intune that needs multiple lifetime:
1. **GenevaInitializationHostLifetime**: to initialize Geneva MDS logging and MDM metrics like account and namespace etc, this needs to happen first before any production logging could be sent to Geneva.
2. **ServiceCrashHostLifetime**: to register TaskScheduler.UnobservedTaskException and AppDomain.CurrentDomain.UnhandledException and report service crash to metrics and logs for trouble-shooting
3. **ServiceFabricContainerAssemblyResolverHostLifetime**: this is needed to support container in Service Fabric, to register AppDomain.CurrentDomain.AssemblyResolve, as some assembly are not in current directory. See [How to containerize your Service Fabric Reliable Services and Reliable Actors](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-services-inside-containers) and SFBinaryLoader(https://github.com/Azure/service-fabric-scripts-and-templates/blob/master/code/SFBinaryLoaderForContainers/SFBinaryLoader.cs)

**Comment by Tratcher** [comment]
Would you always be using all of these, or would you be dynamically deciding which ones to use at startup?

**Comment by chuanboz** [comment]
above 1, 2 we always needs to use, 3 we only use in case it's running inside Service Fabric container, and in that case we need 1, 2, and 3.

Currently we haven't used generic host since it haven't officially released, we're running with asp.net + Service Fabric. but I'm working on switching to generic host. above are scenario are from our current production service. If I switch to generic host, I will need to have this multiple IHostLifetime support.

**Comment by Tratcher** [comment]
That's a code organization question, you're not functionally blocked here. If you always need scenario 1 an 2 then functionally they can be included in the same lifetime. For 3, it could be added via a derived implementation.

**Comment by chuanboz** [comment]
that's a hack workaround and violate Single Responsibility Principle, I want to avoid it if possible.

And you could argue the same for IHostedService, why you will need multiple IHostedService, why not combine them together and in IHost only support one single IHostedService.

for IApplicationLifetime, just one instance is enough as the CancellationToken.Register could be called in multiple places, thus IApplicationLifetime does not needs multiple instance. But that's not the case for IHostLifetime without compromise the design principle.

**Comment by davidfowl** [comment]
Those examples aren't very convincing TBH, can you explain why those are examples of host lifetime implementations?

**Comment by chuanboz** [comment]
do you have any suggested generic host interface to wrap those implementation but not use IHostedLifetime?

**Comment by davidfowl** [comment]
IHostedService has a StartAsync and a StopAsync and it sounds like what you want.

**Comment by davidfowl** [comment]
> That's not really a hosted service, because it's not constantly running after host start. If you compare it with the ConsoleHostLifetime, it is just similar to that, to register some global event or do initialization.

IHostedService isn't about constantly running things. It's about hooking into the existing lifetime of the host.

> If the suggestion is to use IHostedService, why ConsoleHostLifetime is not implemented as a ConsoleHostedService?

HostLifetime controls the lifetime of the host as in it decides why when the host starts up and when the host shuts down.

There are 2 extremely different concerns:
1. I want to **know** when the host starts and when it stops - IHostedService
2. I want to **decide** when the host starts and when it stops - IHostLifetime

**Comment by chuanboz** [comment]
I tried further and still see I could not use the IHostedService for this purpose. 
For production, I'm running with Service Fabric with ServiceFabricHost that replaced regular Host (https://github.com/Azure/service-fabric-aspnetcore/issues/48), but for dev testing I use regular Host, I want to share the same logic for logging initialization and service crash handling.

for Service Fabric, if I model such logic as IHostedService, that mean there could be multiple service instances running in the same process, but I don't need multiple service crash handler, I only need one, and in such case modeling it with IHostLifetime make sense to me from logic perspective.

>That's a code organization question, you're not functionally blocked here. If you always need scenario 1 an 2 then functionally they can be included in the same lifetime
@Tratcher, if I include 1 and 2, then I lost the flexibility in DI, as 1 and 2 has additional parameter coming from DI constructor injection, with including them both I now have to merge these unrelated parameters.

besides the 1, 2, and 3 mentioned, I'd also want to have the built-in ConsoleLifetime experience, if I merge them I now also have a hard dependency on ConsoleLifetime which is against the dependency injection principle.

**Comment by chuanboz** [comment]
here is my attempt to including them all, which really seems ugly IMHO;

```C#
/// <summary>
    /// Ugly workaround needed to support multiple IHostLifetime in Host class
    /// see issue https://github.com/aspnet/Hosting/issues/1401
    /// </summary>
    public class UglyDumpGroundLifetime : IHostLifetime
    {
        private ConsoleLifetime consoleLifetime;
        private ServiceCrashLifetime serviceCrash;
        private UnifiedApiLifetime genevaLogging;

        public UglyDumpGroundLifetime
            (
            ILoggerContext logger, // used for ServiceCrashLifetime
            ICloudEnvironment cloud, IServiceIdentity service, // used for UnifiedApiLifetime
            IOptions<ConsoleLifetimeOptions> options, IHostingEnvironment environment, IApplicationLifetime applicationLifetime // used for ConsoleLifetime
            )
        {
            this.consoleLifetime = new ConsoleLifetime(options, environment, applicationLifetime); // anti-pattern, direct type coupling
            this.serviceCrash = new ServiceCrashLifetime(logger);
            this.genevaLogging = new UnifiedApiLifetime(cloud, service);
        }

        public async Task StopAsync(CancellationToken cancellationToken)
        {
            await this.consoleLifetime.StopAsync(cancellationToken);
            await this.serviceCrash.StopAsync(cancellationToken);
            await this.genevaLogging.StopAsync(cancellationToken);
        }

        public async Task WaitForStartAsync(CancellationToken cancellationToken)
        {
            await this.consoleLifetime.WaitForStartAsync(cancellationToken);
            await this.serviceCrash.WaitForStartAsync(cancellationToken);
            await this.genevaLogging.WaitForStartAsync(cancellationToken);
        }
    }
```

**Comment by davidfowl** [comment]
> for Service Fabric, if I model such logic as IHostedService, that mean there could be multiple service instances running in the same process, but I don't need multiple service crash handler, I only need one, and in such case modeling it with IHostLifetime make sense to me from logic perspective.

I don't understand. Everything you described was about getting notifications, it wasn't about driving the lifetime so IHostedService is a fit for this. Even your code sample kinda shows this.

> That's not really a hosted service, because it's not constantly running after host start. If you compare it with the ConsoleHostLifetime, it is just similar to that, to register some global event or do initialization.

You're looking at the wrong aspect of the code, the console lifetime is controlling when the host stops. Your IHostedService will get notified on start and stop. You can wire up an event in Start and unhook it in Stop. 

I'm just not seeing the problem.

**Comment by chuanboz** [comment]
thanks @davidfowl , that help me understand better of the IHostLifetime vs. IHostedService, and after further thinking I come up with below solution that does not require multiple IHostLifetime but still meet my scenario requirement. Can you help to take a look if that's right thing to do, and whether that's general concept that would benefit generic host.

###

**Comment by Tratcher** [comment]
Where?

**Comment by aspnet-hello** [comment]
We periodically close 'discussion' issues that have not been updated in a long period of time.

We apologize if this causes any inconvenience. We ask that if you are still encountering an issue, please log a new issue with updated information and we will investigate.
